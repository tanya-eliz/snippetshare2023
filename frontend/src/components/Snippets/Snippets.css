.snippets-page{
    padding: 3%;
    padding-top:2.5%;
}

.header-row{
    border-bottom: 2px solid #14213D;
}

.header{
    font-size: 18px !important;
    font-weight: bold !important;
}

img{
    height: 20px;
    vertical-align: middle;
    margin-left: 5px;
}

.cell{
    padding: 12px !important;
}

.pagi-align {
    display: flex;
    align-items: center;
}
  

.table-body{
    margin-top:1%;
}