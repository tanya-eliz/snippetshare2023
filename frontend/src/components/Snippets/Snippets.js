import React, {useEffect, useState} from 'react';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import './Snippets.css';
import ViewIcon from '../../assets/visible.png'
import Pagination from '../../atoms/Pagination/Pagination';
import Button from '../../atoms/Button/Button'
import Stack from '@mui/material/Stack';
import Sort from '../../atoms/Sort/Sort'
import Grid from '@mui/material/Grid';
import { Link } from "react-router-dom";
import useFetch from '../../components/ViewSnippet/useFetch';

const BE_API_URL = process.env.BE_API_URL || 'http://localhost:4000';

export const Snippets = () => {
  const [rowdata, setRowData] = useState([]);
  const [rowsPerPage, setRowsPerPage] = useState(10);
  const [page, setPage] = useState(1);
  const [sort, setSort] = React.useState('Recent Upload');
  const [totalPage, setTotalPage] = React.useState(1);

  useEffect(() => {
    console.log("page:", page)
    let queryString = `?page=${page? page : 1}&limit=${rowsPerPage? rowsPerPage :10}`;
    if ( rowdata.length  && sort === 'Recent Upload') {
      queryString += '&orderBy=createdAt';
    } else if (rowdata.length && sort === 'View Count') {
      queryString += '&orderBy=viewCount';
    }
    customFetch(`${BE_API_URL}/api/snippets` + queryString);
  }, [sort, page, rowdata.length, rowsPerPage]);

  const customFetch = async (url) => {
    try {
      const response = await fetch(url);
      const results = await response.json();
      console.log(results.snippets)
      if (results.snippets) {
        setRowData(results.snippets);
        setTotalPage(results.totalPages);
        setPage(results.currentPage);
      }
    } catch (err) {
      console.log(err);
      alert('Error fetching snippets');
    }
  }

  // Utility function to convert date to local time
  const convertDate = (date) => {
    const newDate = new Date(date)
    console.log(newDate);
    return newDate.toLocaleString('en-SG',{timeZone:'Asia/Singapore'});
  }

  const nextPage = () => {
    setPage(page+1);
  }

  const previousPage = () => {
    setPage(page-1);
  }

  const renderPagination = () => {
    return(
        <div className="pagi-align">
            {
              page>1 && 
              <span>
                <Button label ={'Prev Page'} disabled = {false} variant='pagi' align='center'  onClick={()=>{previousPage()}}/> 
              </span>
            }
            <Pagination currPageNum={page} totalPage={totalPage}/>
            {
              page<totalPage && 
              <span>
                  <Button label ={'Next Page'} disabled = {false} variant='pagi' align='center' onClick={() => {nextPage()}}/> 
              </span>
            }
        </div>
    )
  }

  return (
      <div data-testid = "snippets" className="snippets-page">

        <Grid container spacing={2} sx={{marginBottom: 1.5}}>
          <Grid item xs={10}>
            <h1>View All Snippets</h1>
          </Grid>
          <Grid item xs={2}>
            <Sort sort={sort} setSort={setSort}/>
          </Grid>
        </Grid>
        
        
        <TableContainer component={Paper} sx={{backgroundColor: "#FFFBFF"}}>
          <Table size="small" aria-label="a dense table" className="table-body">
            <TableHead>
              <TableRow className="header-row">
                <TableCell className="header">Title</TableCell>
                <TableCell align="right" className="header">Uploaded Time</TableCell>
                <TableCell align="right" className="header">Views</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              { rowdata.length>0 && rowdata.map((row,index) => (
                <TableRow
                  key={index} 
                  sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                >
                  <TableCell component="th" scope="row">
                    <Link to={`/view-snippet/${row.id}`}>
                        {row.title}
                    </Link>
                  </TableCell>
                  <TableCell className="cell" align="right">{convertDate(row.createdAt)}</TableCell>
                  <TableCell className="cell" align="right">{row.viewCount}<img src={ViewIcon} alt="view"/></TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
          
        </TableContainer>
        
          <div style={{float: 'right', marginTop:'1%'}}>
						<Stack spacing={4} direction="row">
							{renderPagination()}
						</Stack>
          </div>
      </div>

    );
}

export default Snippets