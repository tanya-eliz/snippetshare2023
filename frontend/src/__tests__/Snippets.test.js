/* eslint-disable testing-library/no-container */
/* eslint-disable testing-library/await-async-query */
import {render, screen, cleanup} from "@testing-library/react";
import "@testing-library/jest-dom";
import Snippets from "../components/Snippets/Snippets";

afterEach(() => {
    cleanup();
});

test("should render Snippets component", () => {
    render(<Snippets />);
    const snipElem = screen.getByTestId("snippets");
    expect(snipElem).toBeInTheDocument();
    expect(screen.getByText("View All Snippets")).toBeInTheDocument();
    expect(screen.getByText("Title")).toBeInTheDocument();
    expect(screen.getByText("Uploaded Time")).toBeInTheDocument();
    expect(screen.getByText("Views")).toBeInTheDocument();
});

test("Snippets component matches snapshot", () => {
    const { asFragment } = render(<Snippets />);
    expect(asFragment()).toMatchSnapshot();
});
