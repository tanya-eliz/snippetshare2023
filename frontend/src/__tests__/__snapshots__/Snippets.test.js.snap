// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snippets component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="snippets-page"
    data-testid="snippets"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-yxdi24-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-17p3wh3-MuiGrid-root"
      >
        <h1>
          View All Snippets
        </h1>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
      >
        <div>
          <div
            class="MuiFormControl-root css-7w6d37-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-rrkv15-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="demo-simple-select-autowidth-label"
            >
              <img
                alt="sorter"
                src="sort.png"
                style="margin-right: 5px;"
              />
              Sort By
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="demo-simple-select-autowidth-label demo-simple-select-autowidth"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="demo-simple-select-autowidth"
                role="button"
                tabindex="0"
              >
                Recent Upload
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="Recent Upload"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Sort By
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-r03xzf-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="a dense table"
        class="MuiTable-root table-body css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head header-row css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall header css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Title
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall header css-6qfsqn-MuiTableCell-root"
              scope="col"
            >
              Uploaded Time
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall header css-6qfsqn-MuiTableCell-root"
              scope="col"
            >
              Views
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
    <div
      style="float: right; margin-top: 1%;"
    >
      <div
        class="MuiStack-root css-1xcvc1q-MuiStack-root"
      >
        <div
          class="pagi-align"
        >
          <div>
            <span
              class="MuiTypography-root MuiTypography-tablePagi css-37jdci-MuiTypography-root"
            >
              | Page 1 of 1 |
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
